File: src/commands/core/HelpCommand.ts
/**
 * Copyright (c) 2019-2021 Nino
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Command, CommandMessage, EmbedBuilder } from '../../structures';
import { Constants as ErisConstants } from 'eris';
import { firstUpper } from '@augu/utils';
import CommandService from '../../services/CommandService';
import Permissions from '../../util/Permissions';
import { Inject } from '@augu/lilith';
import Discord from '../../components/Discord';

interface CommandCategories {
  moderation?: Command[];
  settings?: Command[];
  general?: Command[];
}

export default class HelpCommand extends Command {
  private categories!: CommandCategories;
  private parent!: CommandService;

  @Inject
  private discord!: Discord;

  constructor() {
    super({
      description: 'descriptions.help',
      examples: ['help', 'help help', 'help General'],
      cooldown: 2,
      aliases: ['halp', 'h', 'cmds', 'commands'],
      usage: '[cmdOrMod | "usage"]',
      name: 'help'
    });
  }

  run(msg: CommandMessage, [command]: [string]) {
    return command !== undefined ? this.renderDoc(msg, command) : this.renderHelpCommand(msg);
  }

  private async renderHelpCommand(msg: CommandMessage) {
    if (this.categories === undefined) {
      this.categories = {};

      const commands = this.parent.filter(cmd => !cmd.ownerOnly);
      for (let i = 0; i < commands.length; i++) {
        const command = commands[i];
        (this.categories[command.category] ??= []).push(command);
      }
    }

    const prefix = msg.settings.prefixes[Math.floor(Math.random() * msg.settings.prefixes.length)];
    const embed = EmbedBuilder.create()
      .setTitle(msg.locale.translate('commands.help.embed.title', [`${this.discord.client.user.username}#${this.discord.client.user.discriminator}`]))
      .setDescription(msg.locale.translate('commands.help.embed.description', [prefix, this.parent.size]));

    for (const cat in (this.categories as Required<CommandCategories>)) {
      const commands = (this.categories[cat] as Command[]);
      embed.addField(msg.locale.translate(`commands.help.embed.fields.${cat}` as any, [this.categories[cat].length]), commands.map(cmd => `**\`${cmd.name}\`**`).join(', '), false);
    }

    return msg.reply(embed);
  }

  private async renderDoc(msg: CommandMessage, cmdOrMod: string) {
    const command = this.parent.filter(cmd => !cmd.hidden && cmd.name === cmdOrMod || cmd.aliases.includes(cmdOrMod))[0];
    const prefix = msg.settings.prefixes[msg.settings.prefixes.length - 1];

    if (command !== undefined) {
      const description = msg.locale.translate(command.description as any);
      const embed = EmbedBuilder.create()
        .setTitle(msg.locale.translate('commands.help.command.embed.title', [command.name]))
        .setDescription(msg.locale.translate('commands.help.command.embed.description', [description]))
        .addFields(
          [
            {
              name: msg.locale.translate('commands.help.command.embed.fields.syntax'),
              value: `**\`${prefix}${command.format}\`**`,
              inline: false
            },
            {
              name: msg.locale.translate('commands.help.command.embed.fields.category'),
              value: firstUpper(command.category),
              inline: true
            },
            {
              name: msg.locale.translate('commands.help.command.embed.fields.aliases'),
              value: command.aliases.join(', ') || 'No aliases available',
              inline: true
            },
            {
              name: msg.locale.translate('commands.help.command.embed.fields.owner_only'),
              value: command.ownerOnly ? 'Yes' : 'No',
              inline: true
            },
            {
              name: msg.locale.translate('commands.help.command.embed.fields.cooldown'),
              value: `${command.cooldown} Seconds`,
              inline: true
            },
            {
              name: msg.locale.translate('commands.help.command.embed.fields.user_perms'),
              value: Permissions.stringify(command.userPermissions.reduce((acc, curr) => acc | ErisConstants.Permissions[curr], 0n)) || 'None',
              inline: true
            },
            {
              name: msg.locale.translate('commands.help.command.embed.fields.bot_perms'),
              value: Permissions.stringify(command.botPermissions.reduce((acc, curr) => acc | ErisConstants.Permissions[curr], 0n)) || 'None',
              inline: true
            },
            {
              name: msg.locale.translate('commands.help.command.embed.fields.examples'),
              value: command.examples.map(example => `• **${prefix}${example}**`).join('\n') || 'No examples are available.',
              inline: false
            }
          ]
        );

      return msg.reply(embed);
    } else {
      if (cmdOrMod === 'usage') {
        const embed = EmbedBuilder.create()
          .setTitle(msg.locale.translate('commands.help.usage_title'))
          .setDescription(msg.locale.translate('commands.help.usage', [msg.settings.prefixes[0]]));

        return msg.reply(embed);
      }

      const mod = this.parent.filter(cmd => cmd.category.toLowerCase() === cmdOrMod.toLowerCase());
      if (mod.length > 0) {
        const embed = EmbedBuilder.create()
          .setAuthor(msg.locale.translate('commands.help.module.embed.title', [firstUpper(cmdOrMod)]))
          .setDescription(mod.map(command =>
            `**\`${prefix}${command.format}\`** ~  \u200b \u200b**${msg.locale.translate(command.description as any)}**`
          ));

        return msg.reply(embed);
      } else {
        return msg.reply(msg.locale.translate('commands.help.command.not_found', [cmdOrMod]));
      }
    }
  }
}

File: src/main/kotlin/sh/nino/discord/commands/core/HelpCommand.kt
/**
 * Copyright (c) 2019-2021 Nino
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package sh.nino.discord.commands.core

import dev.kord.core.Kord
import sh.nino.discord.core.annotations.Command
import sh.nino.discord.core.command.AbstractCommand
import sh.nino.discord.core.command.CommandCategory
import sh.nino.discord.core.command.CommandHandler
import sh.nino.discord.core.command.CommandMessage
import sh.nino.discord.data.Config
import sh.nino.discord.extensions.asString
import sh.nino.discord.extensions.sort
import sh.nino.discord.extensions.toTitleCase
import sh.nino.discord.utils.Constants
import sh.nino.discord.core.command.Command as BaseCommand

@Command(
    name = "help",
    description = "descriptions.core.help",
    aliases = ["halp", "h", "?", "cmds", "commands"],
    usage = "[cmdOrModule]"
)
class HelpCommand(
    private val handler: CommandHandler,
    private val config: Config,
    private val kord: Kord
): AbstractCommand() {

    override suspend fun run(msg: CommandMessage) = if (msg.args.isEmpty())
        renderHelpCommand(msg)
    else
        renderCommandHelp(msg)

    private suspend fun renderHelpCommand(msg: CommandMessage) {
        // Sort commands by their category
        val categoryMap = mutableMapOf<CommandCategory, MutableList<BaseCommand>>()
        for (cmd in handler.commands.values) {
            // Do not return easter egg or system commands.
            if (cmd.category == CommandCategory.SYSTEM || cmd.category == CommandCategory.EASTER_EGG) continue
            if (!categoryMap.contains(cmd.category))
                categoryMap[cmd.category] = mutableListOf()

            categoryMap[cmd.category]!!.add(cmd)
        }

        val prefix = try {
            msg.settings.prefixes.random()
        } catch (e: NoSuchElementException) {
            config.prefixes.random()
        }

        // Return a pretty embed 🥺
        val self = kord.getSelf()
        msg.replyEmbed {
            title = "${self.username}#${self.discriminator} | Commands"
            description = buildString {
                appendLine(":pencil2: For more documentation, you can type **${prefix}help <cmdOrModule>**, in which, **<cmdOrModule>** is the module or command you want to view.")
                appendLine()
                appendLine("More information and a prettier UI for commands or modules can be viewed on the [website](https://nino.sh/commands).")
                appendLine("There are currently **${handler.commands.size}** available commands.")
            }

            for ((cat, cmds) in categoryMap) {
                field {
                    name = "${cat.emoji} ${cat.category}"
                    value = cmds.joinToString(", ") { "**`${it.name}`**" }
                    inline = false
                }
            }
        }
    }

    private suspend fun renderCommandHelp(msg: CommandMessage) {
        val arg = msg.args.first()
        val prefix = try {
            msg.settings.prefixes.random()
        } catch (e: NoSuchElementException) {
            config.prefixes.random()
        }

        if (arg == "usage") {
            msg.replyEmbed {
                title = "[ Command Usage ]"
                color = Constants.COLOR
                description = buildString {
                    appendLine("> **This is a guide on how to understand the arguments system**")
                    appendLine("> **This is also present on the [website](https://nino.sh/docs/getting-started/syntax)!**")
                    appendLine()
                    appendLine("A command's syntax might look like: `${prefix}help [cmdOrModule]`")
                    appendLine()
                    appendLine("```")
                    appendLine("|-----------------------------------------------------|")
                    appendLine("|                                                     |")
                    appendLine("|      x!    help      [ cmdOrMod   |   \"usage\"]      |")
                    appendLine("|      ^       ^       ^   ^        ^   ^             |")
                    appendLine("|      |       |       |   |        |   |             |")
                    appendLine("|      |       |       |   |        |   |             |")
                    appendLine("|      |       |       |  /         |   |             |")
                    appendLine("|      |       |       | /          |   |             |")
                    appendLine("|      |       |       |/ - name    |   |             |")
                    appendLine("|      |       |       |            |   |             |")
                    appendLine("|    prefix command  param         \"or\" literal       |")
                    appendLine("|-----------------------------------------------------|")
                    appendLine("```")
                    appendLine()
                    appendLine("To break down this thing I shown you above:")
                    appendLine("  - **prefix** refers to the command prefix.")
                    appendLine("  - **command** is the command name or alias.")
                    appendLine("  - **param** is referred to a parameter (or an argument) which can be in `[` or `<`.")
                    appendLine("     - **name** is the param name, nothing much.")
                    appendLine("     - **[]** is an optional argument, so it is not required.")
                    appendLine("     - **<>** is a required argument, it is required to be added.")
                    appendLine()
                    appendLine("In Nino v**2.x**+, we added slash commands to every guild (and some secret ones to some guilds)")
                    appendLine("to making commands, hopefully, efficient as possible.")
                }
            }

            return
        }

        val command = handler.commands.values.firstOrNull {
            (it.name == arg || it.aliases.contains(arg)) && (it.category != CommandCategory.EASTER_EGG && it.category != CommandCategory.SYSTEM)
        }

        if (command != null) {
            msg.replyEmbed {
                title = "[ \uD83D\uDD8C️ Command ${command.name} ]"
                description = msg.locale.translate(command.description)

                field {
                    name = "❯ Syntax"
                    value = "`$prefix${command.name} ${command.usage.trim()}`"
                    inline = false
                }

                field {
                    name = "❯ Category"
                    value = "${command.category.emoji} **${command.category.category}**"
                    inline = true
                }

                field {
                    name = "❯ Alias(es)"
                    value = command.aliases.joinToString(", ").ifEmpty { "None" }
                    inline = true
                }

                field {
                    name = "❯ Examples"
                    value = command.examples.joinToString("\n") { it.replace("{prefix}", prefix) }.ifEmpty { "No examples were provided." }
                    inline = true
                }

                field {
                    name = "❯ Conditions"
                    value = buildString {
                        appendLine("• **Owner Only**: ${if (command.ownerOnly) "Yes" else "No"}")
                    }

                    inline = true
                }

                field {
                    name = "❯ Cooldown"
                    value = "${command.cooldown}s"
                    inline = true
                }

                field {
                    name = "❯ Permissions"
                    value = buildString {
                        appendLine("**User**:")
                        if (command.userPermissions.values.isEmpty()) {
                            appendLine("• **None**")
                        } else {
                            for (perm in command.userPermissions.values.toTypedArray()) {
                                appendLine("• ${perm.asString()}")
                            }
                        }

                        appendLine()
                        appendLine("**Bot**:")
                        if (command.botPermissions.values.isEmpty()) {
                            appendLine("• **None**")
                        } else {
                            for (perm in command.botPermissions.values.toTypedArray()) {
                                appendLine("• ${perm.asString()}")
                            }
                        }
                    }

                    inline = true
                }
            }
        } else {
            val module = handler.commands.values.filter {
                it.category.category.lowercase() == arg.lowercase() && !listOf("system", "easter_egg").contains(arg.lowercase())
            }

            if (module.isNotEmpty()) {
                val propLen = { name: String -> name.length }
                val longestCmdName = propLen(module.sort { a, b -> propLen(b.name) - propLen(a.name) }.first().name)

                msg.replyEmbed {
                    title = "[ Module ${arg.toTitleCase()} ]"
                    description = buildString {
                        for (c in module) {
                            appendLine("`${c.name.padEnd((longestCmdName * 2) - c.name.length, '\u200b')}` | \u200b \u200b**${msg.locale.translate(c.description)}**")
                        }
                    }
                }
            } else {
                msg.reply(":question: Command or module **$arg** was not found.")
                return
            }
        }
    }
}
